src/twisted/trial/test/detests.py

```diff
@@ -186,6 +186,24 @@ def test_timeoutZero(self):
 
     test_timeoutZero.timeout = 0  # type: ignore[attr-defined]
 
+    def test_addCleanupPassDefault(self):
+        # test default timeout for addCleanup functions
+        def cleanup():
+            d = defer.Deferred()
+            reactor.callLater(0, d.callback, "success")
+            return d
+
+        self.addCleanup(cleanup)
+
+    def test_addCleanupTimeout(self):
+        # test that functions added via addCleanup supports timeout
+        def cleanup():
+            return defer.Deferred()
+
+        self.addCleanup(cleanup)
+
+    test_addCleanupTimeout.timeout = 0.1  # type: ignore[attr-defined]
```

Comment:
Ideally, this functionality would be tested against a test double clock so that it could complete nearly instantly (instead of always taking at least 100ms), be robust against inconvenient scheduler decisions maybe the OS, and wouldn't hang forever if the functionality isn't working correctly.

A lot of the existing trial tests are not written in the ideal style.  If you can easily figure out how to rewrite this test, it's worth doing so but if it's not obvious how to do it now then it would also be reasonable to re-visit as a follow-up (not blocking this fix on it).  The more trial maintenance you want to do, the more important it will be to figure out how to write tests in the preferred way, though.